dist/transport.go

```diff
@@ -0,0 +1,186 @@
+package dist
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/containers/image/v5/docker/reference"
+	"github.com/containers/image/v5/image"
+	"github.com/containers/image/v5/transports"
+	"github.com/containers/image/v5/types"
+	"github.com/pkg/errors"
+)
+
+//nolint (gochecknoinits)
+func init() {
+	transports.Register(Transport)
+}
+
+// nolint (gochecknoglobals)
+var Transport = distTransport{}
+
+type distTransport struct{}
+
+func (s distTransport) Name() string {
+	return "dist"
+}
+
+func splitReference(ref string) (fullname, server string, port int, err error) {
+	port = 8080
+	err = nil
+
+	if ref[0] == '/' {
+		fullname = ref[1:]
+		return fullname, "", port, nil
+	}
+
+	fields := strings.SplitN(ref, "/", 2)
+	subFields := strings.Split(fields[0], ":")
+
+	if len(subFields) > 2 {
+		err = fmt.Errorf("bad server:port")
+		return "", "", -1, err
+	}
+
+	server = subFields[0]
+
+	if len(subFields) == 2 {
+		port, err = strconv.Atoi(subFields[1])
+		if err != nil {
+			return "", "", -1, err
+		}
+
+		if port < 1 || port > 65535 {
+			err = fmt.Errorf("bad port %d", port)
+			return "", "", -1, err
+		}
+	}
+
+	fullname = fields[1]
+
+	return fullname, server, port, nil
+}
+
+// NOTE - the transport interface is defined in types/types.go.
+// Valid uris are:
+//    dist:///name1/name2/tag
+//    dist://server/name1/name2/name3/tag
+// The tag can be separated by either / or :
+//    dist://server:port/name1/name2/name3/tag
+//    dist://server:port/name1/name2/name3:tag
+// So the reference passed in here would be e.g.
+//    ///name1/name2/tag
+//    //server:port/name1/name2/tag
```

Comment:
- Overwhelming experience with the Docker references is that it is way too attractive to create them using `Sprintf`, and parse them using `strings.Split`, which _almost always_ handles corner cases incorrectly. If this defines a new reference type of similar importance, it seems _very necessary_ to provide a Go package that does all the parsing, validation and formatting for users, so that users at least have a _chance_ to use something safer. I.e. I feel that something similar to `{docker/distribution,containers/image}/docker/reference` is absolutely necessary.
- The `//` part is pointless visual noise. The values are not URLs (notably no URL escaping), it only leads to extra typing and confused users.
- Why allow the `/tag` vs. `:tag` ambiguity? What’s the end-user benefit?
- Are digest references not supported? Or does “tag” somehow include digests without documentation?